### PR TITLE
Add cloud provider metadata

### PR DESCRIPTION
**Pull request by mikker**
See elastic/apm-agent-ruby#841

### CODE REVIEW

**Comment by estolfo** [comment]
Overall, it looks good but why do we have flat values for some fields while other agents (for ex python) and the documentation has embedded values? I might be missing something in the issue comments...

For example, [this](https://github.com/elastic/apm/blob/master/specs/agents/metadata.md#cloud-provider-metadata) shows `id` and `name` being embedded under `account`. And python uses an embedded value [here](https://github.com/elastic/apm-agent-python/pull/826/files#diff-74c93d0c007ce4c9a6df8204ed4a6798R127). But we use `account_id` and `account_name`.

**Comment by mikker** [comment]
Good catch! I just didn't want to add nested objects when storing them internally. I need to add the nested objects again when turning into json.

**Comment by mikker** [comment]
AWS confirmed:

```
{"metadata":{"service":{"name":"App","environment":"development","version":null,"agent":{"name":"ruby","version":"3.10.1"},"framework":{"name":"Sinatra","version":"2.1.0"},"language":{"name":"ruby","version":"2.7.2"},"runtime":{"name":"ruby","version":"2.7.2"}},"process":{"pid":18346,"title":"/home/ec2-user/.rvm/gems/ruby-2.7.2/bin/rackup","argv":[]},"system":{"hostname":"ip-172-31-4-63.eu-west-1.compute.internal","architecture":"x86_64","platform":"linux","kubernetes":{"namespace":null,"node":{"name":null},"pod":{"name":null,"uid":null}}},"labels":null,"cloud":{"provider":"aws","account":{"id":"907075276364","name":null},"availability_zone":"eu-west-1b","instance":{"id":"i-0f0165f84e8d411be","name":null},"machine":{"type":"t2.micro"},"project":{"id":null,"name":null},"region":"eu-west-1"}}}
```

**Comment by mikker** [comment]
GCE:

```json
{"metadata":{"service":{"name":"App","environment":"development","version":null,"agent":{"name":"ruby","version":"3.10.1"},"framework":{"name":"Sinatra","version":"2.1.0"},"language":{"name":"ruby","version":"2.7.0"},"runtime":{"name":"ruby","version":"2.7.0"}},"process":{"pid":31623,"title":"/home/mikkelmalmberg/.rvm/gems/ruby-2.7.0/bin/rackup","argv":[]},"system":{"hostname":"instance-1","architecture":"x86_64","platform":"linux","kubernetes":{"namespace":null,"node":"{:name=>nil}","pod":"{:name=>nil, :uid=>nil}"}},"labels":null,"cloud":{"provider":"gcp","account":{"id":null,"name":null},"availability_zone":"us-central1-a","instance":{"id":"1558417203220877769","name":"instance-1"},"machine":{"type":"projects/944210798271/machineTypes/e2-micro"},"project":{"id":"944210798271","name":"brnbw-10er"},"region":"us-central1"}}}
```

**Comment by mikker** [comment]
Azure:

```json
{"metadata":{"service":{"name":"App","environment":"development","version":null,"agent":{"name":"ruby","version":"3.10.1"},"framework":{"name":"Sinatra","version":"2.1.0"},"language":{"name":"ruby","version":"2.7.0"},"runtime":{"name":"ruby","version":"2.7.0"}},"process":{"pid":2209,"title":"/home/azureuser/.rvm/gems/ruby-2.7.0/bin/rackup","argv":[]},"system":{"hostname":"apm-test","architecture":"x86_64","platform":"linux","kubernetes":{"namespace":null,"node":"{:name=>nil}","pod":"{:name=>nil, :uid=>nil}"}},"labels":null,"cloud":{"provider":"azure","account":{"id":"81cf7b1d-04ab-440e-9802-5d0fb07d31ce","name":null},"availability_zone":"","instance":{"id":"b06fe3ff-4294-4422-aefa-71bccbbf8a51","name":"apm-test"},"machine":{"type":"Standard_B2s"},"project":{"id":null,"name":"apm-test_group"},"region":"westeurope"}}}
```

File: lib/elastic_apm/util.rb [hunk 1]
```diff
@@ -46,6 +46,8 @@ def self.reverse_merge!(first, *others)
 
     def self.truncate(value, max_length: 1024)
       return unless value
+
+      value = String(value)
```
Comment (estolfo) [inline lib/elastic_apm/util.rb:50]: Just curious why this change is included- was there a bug with this?

File: spec/elastic_apm/metadata/cloud_info_spec.rb [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      end
+
+      context 'auto' do
+        let(:config) { Config.new(cloud_provider: 'auto') }
+
+        context 'timeouts' do
+          it 'tries all three' do
+            WebMock.stub_request(:get, Metadata::CloudInfo::AWS_URI).to_timeout
+            WebMock.stub_request(:get, Metadata::CloudInfo::GCP_URI).to_timeout
+            WebMock.stub_request(:get, Metadata::CloudInfo::AZURE_URI).to_timeout
```
Comment (estolfo) [inline spec/elastic_apm/metadata/cloud_info_spec.rb:153]: cool, didn't know about the `to_timeout` expectation :)

**Review by estolfo — APPROVED** [review]
lgtm!

File: lib/elastic_apm/util.rb (hunk 1 above)
Comment (mikker) [reply lib/elastic_apm/util.rb:50]: Yes, the `account.id` was an integer on, if I remember correctly, Google Cloud Engine.

File: spec/elastic_apm/metadata/cloud_info_spec.rb (hunk 2 above)
Comment (mikker) [reply spec/elastic_apm/metadata/cloud_info_spec.rb:153]: Me neither! Found it when writing this PR.
